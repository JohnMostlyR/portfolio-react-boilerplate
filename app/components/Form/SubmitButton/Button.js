import styled from 'styled-components';
import PropTypes from 'prop-types';

import { iconUrl } from '../HandIcon';
import { theme } from '../theme';
import { BASE_FONT_BOLD } from '../typography';

const { button } = theme;

const Button = styled.button`
  position: relative;
  margin: 0 2rem;
  z-index: 2;
  appearance: none;
  background-color: transparent;
  border: 0;
  filter: ${({ isEnabled }) => (isEnabled ? 'none' : 'grayscale()')};
  padding: 0;
  ${BASE_FONT_BOLD} > span {
    display: inline-block;
    background-color: ${button.backgroundColor};
    border: 1px solid ${button.backgroundColor};
    border-radius: 5px;
    color: ${button.color};
    cursor: ${({ isEnabled }) => (isEnabled ? 'pointer' : 'not-allowed')};
    padding: 8px;
  }

  &::after,
  &::before {
    position: absolute;
    content: '';
    height: 32px;
    top: 50%;
    width: 32px;
    z-index: -1;
    background-image: ${() => `url(${iconUrl})`};
    background-position: center;
    background-repeat: no-repeat;
    background-size: 100%;
    color: inherit;
    transition: transform 0.3s ease-in-out;
  }

  &::after {
    transform: translate3d(-100%, -50%, 0);
  }

  &::before {
    transform: translate3d(0, -50%, 0) rotateY(180deg);
  }

  &:focus,
  &:hover {
    &::after {
      transform: translate3d(25%, -50%, 0);
    }

    &::before {
      transform: translate3d(-125%, -50%, 0) rotateY(180deg);
    }
  }
`;

Button.propTypes = {
  isEnabled: PropTypes.bool,
};

Button.defaultProps = {
  isEnabled: false,
};

export default Button;
