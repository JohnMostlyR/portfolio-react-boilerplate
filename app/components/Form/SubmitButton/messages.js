/*
 * SubmitButton Messages
 *
 * This contains all the text for the SubmitButton component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  idle: {
    id: 'portfolio.components.SubmitButton.idle',
    defaultMessage: 'Send',
  },
  sending: {
    id: 'portfolio.components.SubmitButton.sending',
    defaultMessage: 'Sending...',
  },
  success: {
    id: 'portfolio.components.SubmitButton.success',
    defaultMessage: "Sent! I'll respond as soon as possible",
  },
  error: {
    id: 'portfolio.components.SubmitButton.error',
    defaultMessage: 'Failed. Please try again',
  },
});
