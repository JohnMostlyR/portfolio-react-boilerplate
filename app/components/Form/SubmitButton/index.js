/**
 *
 * SubmitButton
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Button from './Button';
import Label from './Label';
import messages from './messages';

import SVGIcon from '../../../components/SVGIcon';

function setLabel(buttonState) {
  switch (buttonState) {
    case 'sending':
      return <FormattedMessage {...messages.sending} />;
    case 'success':
      return <FormattedMessage {...messages.success} />;
    case 'error':
      return <FormattedMessage {...messages.error} />;
    default:
      return <FormattedMessage {...messages.idle} />;
  }
}

function SubmitButton({ formIsValid, buttonRef, buttonState }) {
  return (
    <Button isEnabled={formIsValid} innerRef={buttonRef}>
      <span>
        <SVGIcon viewBox="0 0 512 512">
          <path d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z" />
        </SVGIcon>
        <Label>{setLabel(buttonState)}</Label>
      </span>
    </Button>
  );
}

SubmitButton.propTypes = {
  formIsValid: PropTypes.bool,
  buttonRef: PropTypes.func,
  buttonState: PropTypes.oneOf(['idle', 'sending', 'success', 'error']),
};

SubmitButton.defaultProps = {
  formIsValid: false,
  buttonState: 'idle',
};

export default SubmitButton;
