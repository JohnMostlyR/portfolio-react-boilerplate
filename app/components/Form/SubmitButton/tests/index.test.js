import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import SubmitButton from '../index';

describe('SubmitButton', () => {
  it('should render and match the snapshot', () => {
    const wrapper = shallow(<SubmitButton />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should adopt the "formIsValid" property', () => {
    const wrapper = shallow(<SubmitButton formIsValid />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('when the `state` changes', () => {
    let wrapper;

    it('should change its appearance when state is `sending`', () => {
      wrapper = shallow(<SubmitButton formIsValid buttonState="sending" />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should change its appearance when state is `success`', () => {
      wrapper = shallow(<SubmitButton formIsValid buttonState="success" />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should change its appearance when state is `error`', () => {
      wrapper = shallow(<SubmitButton formIsValid buttonState="error" />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
