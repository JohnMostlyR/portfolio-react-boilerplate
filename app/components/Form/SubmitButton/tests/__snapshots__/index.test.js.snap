// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitButton should adopt the "formIsValid" property 1`] = `
<Button
  isEnabled={true}
>
  <span>
    <SVGIcon
      viewBox="0 0 512 512"
    >
      <path
        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
      />
    </SVGIcon>
    <Label>
      <FormattedMessage
        defaultMessage="Send"
        id="portfolio.components.SubmitButton.idle"
        values={Object {}}
      />
    </Label>
  </span>
</Button>
`;

exports[`SubmitButton should render and match the snapshot 1`] = `
<Button
  isEnabled={false}
>
  <span>
    <SVGIcon
      viewBox="0 0 512 512"
    >
      <path
        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
      />
    </SVGIcon>
    <Label>
      <FormattedMessage
        defaultMessage="Send"
        id="portfolio.components.SubmitButton.idle"
        values={Object {}}
      />
    </Label>
  </span>
</Button>
`;

exports[`SubmitButton when the \`state\` changes should change its appearance when state is \`error\` 1`] = `
<Button
  isEnabled={true}
>
  <span>
    <SVGIcon
      viewBox="0 0 512 512"
    >
      <path
        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
      />
    </SVGIcon>
    <Label>
      <FormattedMessage
        defaultMessage="Failed. Please try again"
        id="portfolio.components.SubmitButton.error"
        values={Object {}}
      />
    </Label>
  </span>
</Button>
`;

exports[`SubmitButton when the \`state\` changes should change its appearance when state is \`sending\` 1`] = `
<Button
  isEnabled={true}
>
  <span>
    <SVGIcon
      viewBox="0 0 512 512"
    >
      <path
        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
      />
    </SVGIcon>
    <Label>
      <FormattedMessage
        defaultMessage="Sending..."
        id="portfolio.components.SubmitButton.sending"
        values={Object {}}
      />
    </Label>
  </span>
</Button>
`;

exports[`SubmitButton when the \`state\` changes should change its appearance when state is \`success\` 1`] = `
<Button
  isEnabled={true}
>
  <span>
    <SVGIcon
      viewBox="0 0 512 512"
    >
      <path
        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
      />
    </SVGIcon>
    <Label>
      <FormattedMessage
        defaultMessage="Sent! I'll respond as soon as possible"
        id="portfolio.components.SubmitButton.success"
        values={Object {}}
      />
    </Label>
  </span>
</Button>
`;
